presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -416,12 +447,54 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
                     failureDetector,
                     nodeTaskMap,
                     stageSchedulers,
-                    stageLinkages);
+                    stageLinkages,
+                    stagedDependencies,
+                    schedulingQueue);
             stages.addAll(subTree);
 
             SqlStageExecution childStage = subTree.get(0);
             childStagesBuilder.add(childStage);
         }
+
+        for (Reference<StageExecutionPlan> dependency : plan.getDependencies()) {
+            PlanFragmentId fragmentId = dependency.get().getFragment().getId();
+            if (stagedDependencies.contains(fragmentId)) {
+                continue;
+            }
+
+            StageExecutionPlan subStagePlan = dependency.get();
+            List<SqlStageExecution> subTree = createStages(
+                    // dependent stage doesn't produce any output
+                    ((fragment, tasks, noMoreExchangeLocations) -> {}),
+                    nextStageId,
+                    locationFactory,
+                    subStagePlan.withBucketToPartition(bucketToPartition),
+                    nodeScheduler,
+                    remoteTaskFactory,
+                    session,
+                    splitBatchSize,
+                    partitioningCache,
+                    nodePartitioningManager,
+                    queryExecutor,
+                    schedulerExecutor,
+                    failureDetector,
+                    nodeTaskMap,
+                    stageSchedulers,
+                    stageLinkages,
+                    stagedDependencies,
+                    schedulingQueue);
+
+            SqlStageExecution root = subTree.get(0);
+            OutputBuffers outputBuffers = createInitialEmptyOutputBuffers(DISCARDING)
+                    .withBuffer(new OutputBufferId(0), 0)
+                    .withNoMoreBufferIds();
+            root.setOutputBuffers(outputBuffers);
+
+            stages.addAll(subTree);
+            stagedDependencies.add(fragmentId);
```

Comment:
oh.... I might get some sense how this worked... By this you avoid creating the same `SuperStage` multiple times...

Given this...another approach would be separating create `stages` and `SuperStages`, which might be easier to read ? Just a high-level thought.... 

We need to have a name for `SuperStage` ASAP to help communicate :P.  `Section`? `Component`? `Part`? `Fraction`? `Segment` ?